fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/Job.java

```diff
@@ -88,48 +94,33 @@ public List<Rule> getValidRules(GroupExpression groupExpression, List<Rule> cand
 
     public abstract void execute() throws AnalysisException;
 
-    protected Optional<CopyInResult> invokeRewriteRuleWithTrace(Rule rule, Plan before, Group targetGroup) {
+    protected Optional<CopyInResult> invokeRewriteRuleWithTrace(Rule rule, Plan before, Group targetGroup,
+            EventProducer transformTracer) {
         context.onInvokeRule(rule.getRuleType());
-
-        String traceBefore = enableTrace ? getPlanTraceLog() : null;
+        COUNTER_TRACER.log(CounterEvent.of(Memo.getStateId(),
+                CounterType.EXPRESSION_TRANSFORM, targetGroup, targetGroup.getLogicalExpression(), before));
 
         List<Plan> afters = rule.transform(before, context.getCascadesContext());
         Preconditions.checkArgument(afters.size() == 1);
         Plan after = afters.get(0);
-
-        if (after != before) {
-            CopyInResult result = context.getCascadesContext()
-                    .getMemo()
-                    .copyIn(after, targetGroup, rule.isRewrite());
-
-            if ((result.generateNewExpression || result.correspondingExpression.getOwnerGroup() != targetGroup)
-                    && enableTrace) {
-                String traceAfter = getPlanTraceLog();
-                printTraceLog(rule, traceBefore, traceAfter);
-            }
-
-            return Optional.of(result);
+        if (after == before) {
+            return Optional.empty();
         }
 
-        return Optional.empty();
-    }
-
-    protected String getPlanTraceLog() {
-        return context.getCascadesContext()
+        CopyInResult result = context.getCascadesContext()
                 .getMemo()
-                .copyOut(false)
-                .treeString();
-    }
+                .copyIn(after, targetGroup, rule.isRewrite());
 
-    protected String getMemoTraceLog() {
-        return context.getCascadesContext()
-                .getMemo()
-                .getRoot()
-                .treeString();
+        if (result.generateNewExpression || result.correspondingExpression.getOwnerGroup() != targetGroup) {
+            transformTracer.log(TransformEvent.of(targetGroup.getLogicalExpression(), before, afters,
+                            rule.getRuleType()), rule::isRewrite);
+        }
+
+        return Optional.of(result);
     }
 
-    protected void printTraceLog(Rule rule, String traceBefore, String traceAfter) {
-        logger.info("========== {} {} ==========\nbefore:\n{}\n\nafter:\n{}\n",
-                getClass().getSimpleName(), rule.getRuleType(), traceBefore, traceAfter);
+    protected void trace(GroupExpression groupExpression) {
```

Comment:
this function name is too generalize. we should know trace what from function name. and need to add some comment to explain this function